libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -537,31 +540,24 @@ method start*(g: GossipSub) {.async.} =
   ## start pubsub
   ## start long running/repeating procedures
 
-  # interlock start to to avoid overlapping to stops
-  await g.heartbeatLock.acquire()
-
-  # setup the heartbeat interval
-  g.heartbeatRunning = true
-  g.heartbeatFut = g.heartbeat()
-
-  g.heartbeatLock.release()
+  withLock g.heartbeatLock:
+    # setup the heartbeat interval
+    g.heartbeatRunning = true
+    g.heartbeatFut = g.heartbeat()
 
 method stop*(g: GossipSub) {.async.} =
   trace "gossipsub stop"
 
   ## stop pubsub
   ## stop long running tasks
 
-  await g.heartbeatLock.acquire()
-
-  # stop heartbeat interval
   g.heartbeatRunning = false
```

Comment:
This will stop the `g.heartbeat` loop and then lock and `await g.heartbeatFut`, so the `start`, `heartbeat`, `stop` sequences look correct. I'm not 100% sure if we need them at all tho.

Reply:
it is fine cos heartbeat won't run again since you locked and will just wait the sleep of the heartbeat, we might need to cancel the sleep tho cos now it's pretty long... 😄

Reply:
I'm not sure if the locks on the `heartbeat` are adding anything, unless someone tries to accidentally call `heartbeat` twice, it should not be a problem. So I don't see any inherent concurrency issue with it anymore.